coderd/workspaceapps.go

```diff
@@ -156,3 +142,63 @@ func (api *API) workspaceApplicationAuth(rw http.ResponseWriter, r *http.Request
 	u.RawQuery = q.Encode()
 	http.Redirect(rw, r, u.String(), http.StatusSeeOther)
 }
+
+type ValidWorkspaceAppHostnameOpts struct {
+	AllowPrimaryAccessURL bool
+	AllowPrimaryWildcard  bool
+	AllowProxyAccessURL   bool
+	AllowProxyWildcard    bool
+}
+
+// ValidWorkspaceAppHostname checks if the given host is a valid workspace app
+// hostname based on the provided options. It returns a scheme to force on
+// success. If the hostname is not valid or doesn't match, an empty string is
+// returned.
+func (api *API) ValidWorkspaceAppHostname(ctx context.Context, host string, opts ValidWorkspaceAppHostnameOpts) (string, error) {
```

Comment:
This is strange `"", nil` can be returned. It should return an error if nothing matches.

Reply:
A returned error is a 500 response. An empty string means it's not a permitted hostname so you should return a 400.

Reply:
ooof. Unfortunate